File: projects/kmeans_customer_segmentation/clustering.py
import os
import numpy as np
import matplotlib.pyplot as plt

# make it so that the directory can be set
# make it so that there is a message on where the output is produced

# ----------------------------
# 1) Generate synthetic data
# ----------------------------
np.random.seed(42)

n_per_cluster = 150

true_centers = np.array([
    [50,  35],
    [75,  55],
    [105, 60]
], dtype=float)

sd_income = 10
sd_spend  = 8

def make_cluster(center, n, sd_income, sd_spend):
    income = np.random.normal(center[0], sd_income, n)
    spending = np.random.normal(center[1], sd_spend, n)
    return np.column_stack([income, spending])

X = np.vstack([
    make_cluster(true_centers[0], n_per_cluster, sd_income, sd_spend),
    make_cluster(true_centers[1], n_per_cluster, sd_income, sd_spend),
    make_cluster(true_centers[2], n_per_cluster, sd_income, sd_spend),
])

# ----------------------------
# 2) Manual K-means
# ----------------------------
k = 3
max_iter = 30
tolerance = 1e-4

init_idx = np.random.choice(X.shape[0], size=k, replace=False)
centroids = X[init_idx].copy()

# Create output folder
output_dir = "kmeans_frames"
os.makedirs(output_dir, exist_ok=True)

for it in range(1, max_iter + 1):

    # --- Assignment ---
    distances = np.zeros((X.shape[0], k))
    for j in range(k):
        distances[:, j] = np.sum((X - centroids[j])**2, axis=1)

    cluster_assign = np.argmin(distances, axis=1)

    # --- Plot ---
    plt.figure(figsize=(7.5, 5.5))
    plt.scatter(
        X[:, 0], X[:, 1],
        c=cluster_assign,
        s=25,
        alpha=0.9
    )
    plt.scatter(
        centroids[:, 0], centroids[:, 1],
        c=np.arange(k),
        marker="x",
        s=250,
        linewidths=3
    )

    plt.title(f"Iteration {it}")
    plt.xlabel("Annual Income (k$)")
    plt.ylabel("Spending Score (0-100)")
    plt.xlim(15, 140)
    plt.ylim(0, 100)

    # Save image
    filename = f"/home/proud/Desktop/work/AI/kmeans/iter_{it:02d}.png"
    #filename = os.path.join(output_dir, f"iter_{it:02d}.png")
    plt.savefig(filename, dpi=150)
    plt.close()

    # --- Update ---
    new_centroids = np.zeros_like(centroids)
    for j in range(k):
        pts = X[cluster_assign == j]
        if len(pts) == 0:
            new_centroids[j] = X[np.random.randint(0, X.shape[0])]
        else:
            new_centroids[j] = pts.mean(axis=0)

    shift = np.sum((centroids - new_centroids)**2)

    if shift < tolerance:
        print(f"Converged at iteration {it}")
        centroids = new_centroids
        break

    centroids = new_centroids

print("\nFinal centroids:")
print(centroids)

File: projects/customer_segmentation_with_kmeans/clustering.R
# ------------------------------------------------------------
# K-means of synthetic customer data 
# Plotting at each iteration
# ------------------------------------------------------------

set.seed(42)

# 1) Generate synthetic customer data
# not super realistic but easy to understand
n_per_cluster <- 150

# income_k, spend_score
true_centers <- rbind(
  c(50,  35),
  c(75,  55),
  c(105, 60)
)

sd_income <- 10
sd_spend  <- 8

make_cluster <- function(center, n, sd_income, sd_spend) {
  cbind(
    income = rnorm(n, mean = center[1], sd = sd_income),
    spending = rnorm(n, mean = center[2], sd = sd_spend)
  )
}

X <- rbind(
  make_cluster(true_centers[1, ], n_per_cluster, sd_income, sd_spend),
  make_cluster(true_centers[2, ], n_per_cluster, sd_income, sd_spend),
  make_cluster(true_centers[3, ], n_per_cluster, sd_income, sd_spend)
)

df <- data.frame(
  income = X[,1],
  spending = X[,2]
)


# 2) Manual K-means implementation
k <- 3
max_iter <- 30
tolerance <- 1e-4

init_idx <- sample(1:nrow(df), k)
centroids <- as.matrix(df[init_idx, ])

for (iter in 1:max_iter) {
  
  # --- Assignment ---
  # squared euclidean distances
  distances <- sapply(seq_len(k), function(j) {
    income_diff   <- df$income   - centroids[j, 1]
    spending_diff <- df$spending - centroids[j, 2]
    
    income_diff^2 + spending_diff^2
  })
  
  cluster_assign <- apply(distances, 1, which.min)
  
  # --- Plot ---
  plot(
    df$income, df$spending,
    col = cluster_assign,
    pch = 19, cex = 0.8,
    main = paste("Iteration", iter),
    xlab = "Annual Income (k$)",
    ylab = "Spending Score (0-100)",
    xlim = c(15, 140),
    ylim = c(0, 100)
  )
  
  points(
    centroids[,1], centroids[,2],
    pch = 4, cex = 2.5, lwd = 3,
    col = 1:k
  )
  
  Sys.sleep(1)
  
  # --- Update ---
  new_centroids <- matrix(0, nrow = k, ncol = 2)
  
  for (j in 1:k) {
    cluster_points <- df[cluster_assign == j, ]
    new_centroids[j, ] <- colMeans(cluster_points)
  }
  
  shift <- sum((centroids - new_centroids)^2)
  
  if (shift < tolerance) {
    cat("Converged at iteration", iter, "\n")
    break
  }
  
  centroids <- new_centroids
}

cat("\nFinal centroids:\n")
print(centroids)